src/libsyntax/ast.rs

```diff
@@ -989,32 +1077,56 @@ impl Expr {
         }
     }
 
+    /// Attempts to reparse as `Ty` (for diagnostic purposes).
     pub(super) fn to_ty(&self) -> Option<P<Ty>> {
+        // Attempt to reparse expression.
+        // Since we're working on the AST level, it's possible to not
+        // know the result type of expression yet.
         let kind = match &self.kind {
+            // Trivial conversions.
             ExprKind::Path(qself, path) => TyKind::Path(qself.clone(), path.clone()),
             ExprKind::Mac(mac) => TyKind::Mac(mac.clone()),
+
+            // To convert parentheses to type, convert inner expression to type and
+            // wrap it with `TyKind::Paren`.
             ExprKind::Paren(expr) => expr.to_ty().map(TyKind::Paren)?,
+
+            // Type of `AddrOf` expression (`&some_expr`) is the reference (`TyKind::Rptr`)
+            // to the type of `some_expr`.
             ExprKind::AddrOf(mutbl, expr) => expr
                 .to_ty()
                 .map(|ty| TyKind::Rptr(None, MutTy { ty, mutbl: *mutbl }))?,
+
+            // Type of repeat expression (`[1; 5]`) is type of inner expression
+            // wrapped into `TyKind::Array`.
             ExprKind::Repeat(expr, expr_len) => {
                 expr.to_ty().map(|ty| TyKind::Array(ty, expr_len.clone()))?
             }
+
+            // Only assume type of array at this stage if it contains one element.
```

Comment:
Some comments here feel like
```rust
// Prints a string.
print(string);
```
Perhaps comments repeating what the code says exactly can be removed.